src/ReactiveUI/Routing/RoutableViewModelMixin.cs

```diff
@@ -115,14 +113,21 @@ public static IObservable<Unit> WhenNavigatingFromObservable(this IRoutableViewM
 
             var router = item.HostScreen.Router;
             var navigationStackChanged = router.NavigationChanged?.CountChanged();
-            bool StackIsCleared(Change<IRoutableViewModel> change) => change.Reason == ListChangeReason.Clear;
-            bool ThisViewModelIsRemoved(Change<IRoutableViewModel> change) => NavigationStackRemovalOperations.Contains(change.Reason) && change.Item.Current == item;
-            var itemRemoved = navigationStackChanged.Where(x => x.Any(change => StackIsCleared(change) || ThisViewModelIsRemoved(change)));
+            var itemRemoved = navigationStackChanged.Where(x => WasItemRemoved(x, item));
             var viewModelsChanged = navigationStackChanged.Scan(new IRoutableViewModel[2], (previous, current) => new[] { previous[1], router.GetCurrentViewModel() });
+
             return viewModelsChanged
                 .Where(x => x[0] == item)
                 .Select(_ => Unit.Default)
                 .TakeUntil(itemRemoved);
         }
+
+        private static bool WasItemRemoved(IChangeSet<IRoutableViewModel> changeSet, IRoutableViewModel item)
+        {
+            bool StackIsCleared(Change<IRoutableViewModel> change) => change.Reason == ListChangeReason.Clear;
```

Comment:
Nitpicking wondering if the two local functions are needed, they will have a function invocation cost to them.

Reply:
There we go. Removed.